tests/cupyx_tests/jit_tests/test_raw.py

```diff
@@ -359,6 +359,46 @@ def f(x, y, z):
         assert bool((y == 15).all())
         assert bool((z == 28).all())
 
+    def test_pass_by_reference(self):
+        @jit.rawkernel(device=True)
+        def func(a, b, c):
+            a += 1
+            b = 1  # noqa: F841
+            c -= 1
+
+        @jit.rawkernel()
+        def f(x, y, z):
+            tid = jit.grid(1)
+
+            a = 4
+            b = 4
+            c = 4
+
+            if (tid == 1):
+                func(jit.ref(a), jit.ref(b), c)
+            else:
+                # func is already defined,
+                # a, b will still be passed by reference
+                # c will still be passed by value
+                func(a, b, jit.ref(c))
+
+            # still test pass-by-reference with c
+            d = jit.ref(c)
+            e = 7
+            d += e
+
+            x[tid] = a
+            y[tid] = b
+            z[tid] = c
```

Comment:
I think this style of syntax is confusing. In this test code, mutating argument `b` in func is different from Python's behavior.

How about introducing `jit.pointer_of` like as follows?
```py
@jit.rawkernel(device=True)
def func(a, b, c):
    a.value += 1
    b.value = 1
    c -= 1

@jit.rawkernel()
def f(x, y, z):
    ...
    func(jit.pointer_of(a), jit.pointer_of(b), c)
    ...
```

Reply:
FWIW, Cython uses custom operators `address` (for `ref`/`pointer_of` here) and `dereference` (for `.value` in the above snippet). I agree we need two operators to avoid ambiguity.
https://cython.readthedocs.io/en/latest/src/userguide/wrapping_CPlusPlus.html#c-operators-not-compatible-with-python-syntax

Reply:
I originally thought it would be fine because you can't reassign a reference variable once created (at least not without the new keyword), but I do agree it diverges significantly from python syntax. I'll rework it